File: src/main/java/view25d/view/TurtleViewOptions.java
package view25d.view;

import java.awt.GridLayout;

import javax.swing.ButtonGroup;
import javax.swing.JCheckBox;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

public class TurtleViewOptions extends JPanel implements ChangeListener {
	private static final long serialVersionUID = 1L;

	private double stemThickness = 0.0;
	TurtleView myWindow;
	JCheckBox color, size, shape, pcolor;
  private JRadioButton threeDButton, twoDButton;


	public TurtleViewOptions( TurtleView parent, boolean co, boolean sz, boolean sh, boolean pco, boolean threeD, boolean twoD  ) {
		super();

		this.myWindow = parent;

		this.setLayout(new GridLayout(2, 5));
		color = new JCheckBox("Color", co);
		size = new JCheckBox("Size", sz);
		shape = new JCheckBox("Shape", sh);
		pcolor = new JCheckBox("PColor", pco);

		color.addChangeListener(this);
		size.addChangeListener(this);
		shape.addChangeListener(this);
		pcolor.addChangeListener(this);

		this.add( new JLabel("  View Options:") );
		this.add(color);
		this.add(size);
		this.add(shape);
		this.add(pcolor);

    threeDButton = new JRadioButton("xyz");
		threeDButton.setActionCommand("xyz");
		if (threeD) { threeDButton.setSelected(true); }

		twoDButton = new JRadioButton("xy-plane");
		twoDButton.setActionCommand("xy-plane");
		if (twoD) { twoDButton.setSelected(true); }

		ButtonGroup group2 = new ButtonGroup();
		group2.add(threeDButton);
		group2.add(twoDButton);

		threeDButton.addChangeListener(this);
		twoDButton.addChangeListener(this);

		this.add( new JLabel( "  Link Options:") );
		this.add( threeDButton );
		this.add( twoDButton );
		this.add( new JLabel( "") );
		this.add( new JLabel( "") );
	}

	public boolean showColor() { return color.isSelected(); }
	public boolean showSize() { return size.isSelected(); }
	public boolean showShape() { return shape.isSelected(); }
	public boolean usePColor() { return pcolor.isSelected(); }
	public double getStemThickness() { return stemThickness; }
	public void setStemThickness(double thickness) {
		if (thickness > 0) { stemThickness = thickness;}
		else { stemThickness = 0; }
	}
	public boolean linksAreThreeD() { return threeDButton.isSelected(); }
	public boolean linksAreTwoD() { return twoDButton.isSelected(); }
	public void setLinksDisplayMode(boolean linksInXYPlane) {
		if (linksInXYPlane) {
			twoDButton.setSelected(true);
		} else {
			threeDButton.setSelected(true);
		}
	}

	@Override
	public void stateChanged(ChangeEvent arg0) {
		//System.err.println("state.  now usepcolor selection is: " + usePColor() );
		//System.err.println("source = " + arg0.getSource());
		myWindow.refresh();
	}

}

File: src/main/scala/view/TurtleViewOptions.scala
package view25d.view

import java.awt.{ GridBagConstraints, GridBagLayout, Insets }
import javax.swing.{ ButtonGroup, JLabel, JPanel }

import org.nlogo.swing.{ CheckBox, RadioButton, Transparent }
import org.nlogo.theme.{ InterfaceColors, ThemeSync }

class TurtleViewOptions(window: TurtleView, co: Boolean, sz: Boolean, sh: Boolean, pco: Boolean, threeD: Boolean,
                        twoD: Boolean) extends JPanel(new GridBagLayout) with ThemeSync {

	private var stemThickness = 0.0

  private val color = new CheckBox("Color", (selected) => window.refresh()) {
    if (co)
      setSelected(true)
  }

  private val sizeBox = new CheckBox("Size", (selected) => window.refresh()) {
    if (sz)
      setSelected(true)
  }

  private val shape = new CheckBox("Shape", (selected) => window.refresh()) {
    if (sh)
      setSelected(true)
  }

  private val pcolor = new CheckBox("PColor", (selected) => window.refresh()) {
    if (pco)
      setSelected(true)
  }

  private val threeDButton = new RadioButton("xyz", window.refresh _) {
    if (threeD)
      setSelected(true)
  }

  private val twoDButton = new RadioButton("xy-plane", window.refresh _) {
    if (twoD)
      setSelected(true)
  }

  private val viewLabel = new JLabel("View Options:")
  private val linkLabel = new JLabel("Link Options:")

  locally {
    val group = new ButtonGroup

    group.add(threeDButton)
    group.add(twoDButton)

    val c = new GridBagConstraints

    c.gridy = 0
    c.anchor = GridBagConstraints.WEST
    c.insets = new Insets(6, 6, 6, 6)

    add(viewLabel, c)

    c.insets = new Insets(6, 0, 6, 6)

    add(color, c)
    add(sizeBox, c)
    add(shape, c)
    add(pcolor, c)

    c.weightx = 1

    // invisible buffer to keep everything on the left
    add(new JPanel with Transparent, c)

    c.gridy = 1
    c.weightx = 0
    c.insets = new Insets(0, 6, 6, 6)

    add(linkLabel, c)

    c.insets = new Insets(0, 0, 6, 6)

    add(threeDButton, c)
    add(twoDButton, c)
  }

  def showColor: Boolean =
    color.isSelected

  def showSize: Boolean =
    sizeBox.isSelected

  def showShape: Boolean =
    shape.isSelected

  def usePColor: Boolean =
    pcolor.isSelected

  def getStemThickness: Double =
    stemThickness

  def setStemThickness(value: Double): Unit = {
    stemThickness = value.max(0)
  }

  def linksAreThreeD: Boolean =
    threeDButton.isSelected

  def linksAreTwoD: Boolean =
    twoDButton.isSelected

  def setLinksDisplayMode(linksInXYPlane: Boolean): Unit = {
    if (linksInXYPlane)
      twoDButton.setSelected(true)
    else
      threeDButton.setSelected(true)
  }

  def syncTheme(): Unit = {
    setBackground(InterfaceColors.toolbarBackground())

    viewLabel.setForeground(InterfaceColors.toolbarText())
    linkLabel.setForeground(InterfaceColors.toolbarText())
    color.setForeground(InterfaceColors.toolbarText())
    sizeBox.setForeground(InterfaceColors.toolbarText())
    shape.setForeground(InterfaceColors.toolbarText())
    pcolor.setForeground(InterfaceColors.toolbarText())

    threeDButton.syncTheme()
    twoDButton.syncTheme()
  }
}
